lua/nvim-treesitter/indent.lua

```diff
@@ -0,0 +1,96 @@
+local parsers = require'nvim-treesitter.parsers'
+local queries = require'nvim-treesitter.query'
+
+local M = {}
+
+local function get_node_at_line(root, lnum)
+  for node in root:iter_children() do
+    local srow, _, erow = node:range()
+    if srow == lnum then return node end
+
+    if node:child_count() > 0 and srow < lnum and lnum <= erow then
+      return get_node_at_line(node, lnum)
+    end
+  end
+
+  local wrapper = root:descendant_for_range(lnum, 0, lnum, -1)
+  local child = wrapper:child(0)
+  return child or wrapper
+end
+
+local function update_indents(bufnr)
+  local indents = queries.get_capture_matches(bufnr, '@indent', 'indents') or {}
+  local branches = queries.get_capture_matches(bufnr, '@branch', 'indents') or {}
+
+  local indent_nodes = {}
+  for _, match in ipairs(indents) do
+    if match.node then
+      table.insert(indent_nodes, match.node)
+    end
+  end
+
+  local branch_nodes = {}
+  for _, match in ipairs(branches) do
+    if match.node then
+      table.insert(branch_nodes, match.node)
+    end
+  end
+
+  return { indents = indent_nodes, branches = branch_nodes }
+end
+
+local get_indents = (function()
+  local cache = {}
+
+  return function(bufnr)
+    if cache[bufnr] then
+      return cache[bufnr]
+    end
+
+    vim.api.nvim_buf_attach(bufnr, true, {
+        on_bytes = function() cache[bufnr] = update_indents(bufnr) end,
+        on_detach = function() cache[bufnr] = nil end
+      })
+
+    cache[bufnr] = update_indents(bufnr)
+    return cache[bufnr]
+  end
+end)()
+
+function M.get_indent(lnum)
+  local parser = parsers.get_parser()
+  if not parser or not lnum then return -1 end
+
+  local node = get_node_at_line(parser:parse():root(), lnum-1)
+  local indent_queries = get_indents(vim.api.nvim_get_current_buf())
+  local indents = indent_queries.indents
+  local branches = indent_queries.branches
+  if not indents then return 0 end
+
+  while node and vim.tbl_contains(branches, node) do
```

Comment:
`vim.tbl_contains` is not the ideal solution for lookup. Why not building up a hash table `braches[node] = true` for look-up if `branches[node]`

Reply:
i could, but this is the default implementation for locals so i just followed the same pattern

Reply:
building up the dictionaries like the locals do take a lot of time. probably the fastest option is to create a dictionary while iterating of the raw queries.

Reply:
how well yeah it's damn complex indeed